torchtext/datasets/imdb.py

```diff
@@ -23,16 +28,25 @@
 @_add_docstring_header(num_lines=NUM_LINES, num_classes=2)
 @_create_dataset_directory(dataset_name=DATASET_NAME)
 @_wrap_split_argument(('train', 'test'))
-def IMDB(root, split):
-    def generate_imdb_data(key, extracted_files):
-        for fname in extracted_files:
-            *_, split, label, file = Path(fname).parts
-
-            if key == split and (label in ['pos', 'neg']):
-                with io.open(fname, encoding="utf8") as f:
-                    yield label, f.read()
-    dataset_tar = download_from_url(URL, root=root,
-                                    hash_value=MD5, hash_type='md5')
-    extracted_files = extract_archive(dataset_tar)
-    iterator = generate_imdb_data(split, extracted_files)
-    return _RawTextIterableDataset(DATASET_NAME, NUM_LINES[split], iterator)
+def IMDB(root: str, split: Union[Tuple[str], str]):
+    if not is_module_available("torchdata"):
+        raise ModuleNotFoundError("Package `torchdata` not found. Please install following instructions at `https://github.com/pytorch/data`")
+
+    url_dp = IterableWrapper([URL])
+
+    cache_dp = url_dp.on_disk_cache(
+        filepath_fn=lambda x: os.path.join(root, _PATH),
+        hash_dict={os.path.join(root, _PATH): MD5}, hash_type="md5"
+    )
+    cache_dp = HttpReader(cache_dp).end_caching(mode="wb", same_filepath_fn=True)
+    cache_dp = FileOpener(cache_dp, mode="b")
+
+    extracted_files = cache_dp.read_from_tar()
+
+    def filter_imdb_data(key, fname):
+        *_, split, label, file = Path(fname).parts
+        return key == split and (label in ['pos', 'neg'])
+
+    filter_extracted_files = extracted_files.filter(lambda t: filter_imdb_data(split, t[0]))
+
+    return filter_extracted_files.readlines(decode=True).map(lambda t: (Path(t[0]).parts[-2], t[1]))
```

Comment:
Can we also add double caching similar to what we're doing in the [YelpFullReview dataset PR](https://github.com/pytorch/text/pull/1529/files/1ff59b781c5ee72529f742382e0ad3d0b7871582..61cad246874ba2ac7f6e03f8ab93838f89287540#diff-8056dddf0ce233418c7df18c6c8e9d71e5f81498c0c5b2fe9dcfb793144e1be8R43-R54)?